File: algorithms/greedy.py
from algorithms import algorithm
from utils.reader import GraphReader
import heapq

class Greedy(algorithm.Algorithm):
    def __init__(self, graph, start_node, end_node, cities):
        super().__init__(graph, start_node, end_node)
        self.cities = cities
        self.came_from = {start_node: None}
        self.pq = [(self._heuristic(start_node, end_node), start_node)]
        self.closed_set = set()

    def _heuristic(self, node1, node2):
        return GraphReader.distance(self.cities[node1], self.cities[node2])

    def solve(self):
        self.frontier.add(self.start_node)
        yield self.frontier, self.closed_set, self.start_node, [self.start_node]

        while self.pq:
            _, current_node = heapq.heappop(self.pq)
            self.frontier.discard(current_node)
            
            if current_node in self.closed_set:
                continue

            self.closed_set.add(current_node)

            if current_node == self.end_node:
                yield self.frontier, self.closed_set, current_node, self.reconstruct_path()
                return

            for neighbor in range(len(self.graph[current_node])):
                if self.graph[current_node][neighbor] > 0: # Connected
                    if neighbor not in self.closed_set:
                        self.came_from[neighbor] = current_node
                        heapq.heappush(self.pq, (self._heuristic(neighbor, self.end_node), neighbor))
                        self.frontier.add(neighbor)
            
            yield self.frontier, self.closed_set, current_node, self.reconstruct_path(current_node)

    def reconstruct_path(self, target_node=None):
        if target_node is None:
            target_node = self.end_node
        path = []
        current_node = target_node
        while current_node is not None:
            path.append(current_node)
            if current_node == self.start_node:
                break
            current_node = self.came_from.get(current_node)
        path.reverse()
        return path

File: src/app/core/algorithms/greedy.ts
import { Algorithm, NodeState } from './algorithm';
import { City } from '../models/city';

export class Greedy extends Algorithm {
  private closedSet: Set<number> = new Set();
  private frontierSet: Set<number> = new Set();
  private pq: { hScore: number, node: number }[] = [];
  private cameFrom: Map<number, number> = new Map();
  private cities: City[];

  constructor(graph: number[][], startNode: number, endNode: number, cities: City[]) {
    super(graph, startNode, endNode);
    this.cities = cities;
    this.pq.push({ hScore: this.heuristic(startNode, endNode), node: startNode });
  }

  private heuristic(node1: number, node2: number): number {
    const c1 = this.cities[node1];
    const c2 = this.cities[node2];
    return Math.sqrt(Math.pow(c1.x - c2.x, 2) + Math.pow(c1.y - c2.y, 2));
  }

  *solve(): Generator<NodeState, void, unknown> {
    this.frontierSet.add(this.startNode);
    yield {
      frontier: new Set(this.frontierSet),
      visited: new Set(this.closedSet),
      currentNode: this.startNode,
      path: [this.startNode]
    };

    while (this.pq.length > 0) {
      this.pq.sort((a, b) => a.hScore - b.hScore);
      const current = this.pq.shift()!;
      const currentNode = current.node;

      this.frontierSet.delete(currentNode);

      if (this.closedSet.has(currentNode)) {
        continue;
      }

      this.closedSet.add(currentNode);

      if (currentNode === this.endNode) {
        yield {
          frontier: new Set(this.frontierSet),
          visited: new Set(this.closedSet),
          currentNode: currentNode,
          path: this.reconstructPath()
        };
        return;
      }

      for (let neighbor = 0; neighbor < this.graph[currentNode].length; neighbor++) {
        const weight = this.graph[currentNode][neighbor];
        if (weight > 0) { // Connected
          if (!this.closedSet.has(neighbor)) {
            this.cameFrom.set(neighbor, currentNode);
            this.pq.push({ hScore: this.heuristic(neighbor, this.endNode), node: neighbor });
            this.frontierSet.add(neighbor);
          }
        }
      }

      yield {
        frontier: new Set(this.frontierSet),
        visited: new Set(this.closedSet),
        currentNode: currentNode,
        path: this.reconstructPath(currentNode)
      };
    }
  }

  private reconstructPath(targetNode: number | null = null): number[] {
    let current = targetNode !== null ? targetNode : this.endNode;
    const path: number[] = [];
    while (this.cameFrom.has(current)) {
      path.push(current);
      if (current === this.startNode) break;
      current = this.cameFrom.get(current)!;
    }
    path.push(this.startNode);
    // Remove duplicates if start was pushed twice
    return Array.from(new Set(path)).reverse();
  }
}
